### PR TITLE
New datagen API

### PR DESCRIPTION
**Pull request by robertbastian**
Supersedes #3142

This gets rid of a lot of the current API surface (it's kept around `doc(hidden)` for semver). Instead of calling a top-level `datagen` function with `Out` parameters, clients now construct a `DatagenProvider` with `SourceData` and `Options`, and then call `export` on that. `export` accepts an `impl DataExporter`. This will make it easier to exclude exporter crates (`icu_provider_fs`, `icu_provider_blob`, future `icu_provider_bake`), as the new API can be built without them (#3365).

The new API doesn't support multiple exporters at once. The reason for that is that it's a very niche use case, it can be worked around by defining a forking data exporter, and it's still available through the old API. `make-testdata` still uses the old API, whereas `icu4x-datagen` uses the new one, this way we keep full coverage. When we remove the old API, we can move the forking exporter from datagen to `make-testdata`.

#3365

### CODE REVIEW

**Comment by Manishearth** [comment]
This is going to take some time to land so it's probably fine but I'd like to request we don't land this too soon (not this week, at least). @sffc has previously expressed that he considers this repo to still be in "patch 1.2" state so we shouldn't land major changes that we're not okay with having sneak out through a patch release.

(Plus I haven't finished the Google3 datagen import yet. Hopefully this week.)

**Comment by robertbastian** [comment]
Sure, I'd appreciate feedback already though.

**Comment by Manishearth** [comment]
Oh absolutely, I plan to go through this sooner than that :smile:

**Review by Manishearth — COMMENTED** [review]
partial review

**Review by Manishearth — COMMENTED** [review]
reviewed the whole thing. No blocking comments I'd say

**Comment by Manishearth** [comment]
I am now comfortable landing this. capi may still have a patch release to fix the ssize_t issue if Makoto needs, but I think they use ICU4X from Github.

File: ffi/diplomat/Cargo.toml
Comment (sffc) [inline ffi/diplomat/Cargo.toml:36]: Nit: Unrelated change

File: provider/blob/src/export/mod.rs
Comment (sffc) [inline provider/blob/src/export/mod.rs:11]: Issue: This example code isn't being run. Please add back the \`\`\`

File: provider/datagen/Cargo.toml
Comment (sffc) [inline provider/datagen/Cargo.toml:51]: Issue: Can we avoid the required deserialize_json dependency?

File: provider/datagen/src/baked_exporter.rs
Comment (sffc) [inline provider/datagen/src/baked_exporter.rs:42]: Issue: Fix missing semicolon and indentation.

Request (optional): It would be nice if this example could compile. You could potentially do this by pointing at a path that contains real hello-world bake data.

File: provider/datagen/src/baked_exporter.rs
Comment (sffc) [inline provider/datagen/src/baked_exporter.rs:82]: Suggestion (optional): when two booleans depend on each other like this, it sometimes means that you should be using a 3-element enum instead

```rust
pub enum CrateMode {
  MetaCrate,
  SeparateCratesNoFeatures,
  SeparateCratesWithFeatures,
}
```

File: provider/datagen/src/bin/datagen/args.rs
Comment (sffc) [inline provider/datagen/src/bin/datagen/args.rs:305]: Nit, here and elsewhere: Always say "Cargo feature", not just "feature"

File: provider/datagen/src/bin/datagen/args.rs
Comment (sffc) [inline provider/datagen/src/bin/datagen/args.rs:352]: Question: how does this `/stdout` work? Is it portable?

File: provider/datagen/src/lib.rs
Comment (sffc) [inline provider/datagen/src/lib.rs:110]: Question: why the special case for docs? If this is a non exhaustive struct, it should be written as such in code. We can always document something as being non exhaustive, but the only purpose of the non_exhaustive annotation is to actually enforce that.

File: provider/datagen/src/transform/cldr/displaynames/language.rs
Comment (sffc) [inline provider/datagen/src/transform/cldr/displaynames/language.rs:64]: Discussion/Issue: I'm not really a fan of pulling this into each and every transformer. It seems very much prone to error.

I think I still prefer the design where the list of locales (or locale config) is provided directly to the `export` function instead of being in the datagen constructor.

File: provider/datagen/src/transform/segmenter/mod.rs
Comment (sffc) [inline provider/datagen/src/transform/segmenter/mod.rs:830]: Yes, this is the big case where we likely want to return data for locales that aren't necesarily in the user specified set.

We probably should have an additional LSTM-specific flag to allow you to choose which models to build (for which languages, grapheme/codepoint, training fidelity, etc).

**Review by sffc — CHANGES_REQUESTED** [review]
Like the design overall. Much cleaner. Mostly small nits with a couple discussion questions

File: provider/datagen/Cargo.toml
Comment (robertbastian) [reply provider/datagen/Cargo.toml:51]: `From<serde_json::Error> for DataError` is enabled by that, which would introduce a lot of boilerplate in this crate if we didn't have it.

File: provider/datagen/src/baked_exporter.rs
Comment (robertbastian) [reply provider/datagen/src/baked_exporter.rs:82]: This is pre-existing code (`crate::BakedOptions`) and is re-exported as that for stability.

Also, you wrote this 😀

File: provider/datagen/src/bin/datagen/args.rs
Comment (robertbastian) [reply provider/datagen/src/bin/datagen/args.rs:352]: It's just a magic constant so that we don't have to introduce an enum between stdout and a file path in the `Config`. I don't think stdout is a useful enough feature to complicate the config schema, and it can be achieved with setting the out path to `/dev/stdout` on Unix at least.

File: provider/datagen/src/lib.rs
Comment (robertbastian) [reply provider/datagen/src/lib.rs:110]: It should be an exhaustive struct, but we cannot make it exhaustive because that would be semver breaking (it's currently possible to struct-construct this). We can make it *look* like an exhaustive struct in Rustdoc while not breaking existing code.

We could also mark the field `#[doc(hidden)]`, then the docs would show `/* private fields */` but not `#[non_exhaustive]`.

File: provider/datagen/src/transform/cldr/displaynames/language.rs
Comment (robertbastian) [reply provider/datagen/src/transform/cldr/displaynames/language.rs:64]: If `export` decided which locales to export, any key-specific logic (like for segmentation) would need to be in `export`. I think that's bad design because it would literally be an `if key == Marker::KEY { // choose locales }` for any number of keys, whereas with this design the logic for each key is in its provider implementation.

Comment (robertbastian) [reply provider/datagen/src/transform/cldr/displaynames/language.rs:64]: Whether the locale config is provided to the constructor or `export` is independent of whether the actual logic is in `export` or the providers. Putting it in the constructor is just easier because passing it to `supported_locales` would require changing the public `IterableDataProvider` trait (or using internal mutability on `self` before calling it). I'm somewhat neutral as to where to provide it (I think `export` makes sense as well), but I think evaluating it in the provider impls is much cleaner.

File: provider/datagen/src/transform/segmenter/mod.rs
Comment (robertbastian) [reply provider/datagen/src/transform/segmenter/mod.rs:830]: I guess I'll keep this in for now to preserve behaviour? 

Do we an issue for segmenter datagen locale selection? I couldn't find an open one.

File: provider/datagen/Cargo.toml
Comment (sffc) [reply provider/datagen/Cargo.toml:51]: `deserialize_json` is one of a handful of "mutating" Cargo features in icu4x where enabling the feature is not just a no-op exposing a few more APIs; it increases the size of all BufferProvider impls. Therefore it is important that we are able to disable it. Having `icu_datagen` in your build should not force `deserialize_json` (note: Bazel and GN use global feature resolution).

Comment (robertbastian) [reply provider/datagen/Cargo.toml:51]: Well currently datagen enables it through `icu_provider_fs`, so this PR is not a regression in that sense. I agree it would be nice to be able to disable it, but I don't see this as blocking.

File: provider/datagen/src/bin/datagen/args.rs
Comment (robertbastian) [reply provider/datagen/src/bin/datagen/args.rs:352]: I want to keep this an undocumented implementation detail just to pipe the CLI functionality through `Config`.

File: provider/datagen/src/transform/segmenter/mod.rs
Comment (sffc) [reply provider/datagen/src/transform/segmenter/mod.rs:826]: Created https://github.com/unicode-org/icu4x/issues/3408

File: provider/datagen/src/transform/cldr/displaynames/language.rs
Comment (sffc) [reply provider/datagen/src/transform/cldr/displaynames/language.rs:64]: This seems not quite right to me but it's fine to land for now, and we can discuss further. https://github.com/unicode-org/icu4x/issues/3409

**Review by sffc — DISMISSED** [review]

**Review by Manishearth — APPROVED** [review]

**Review by Manishearth — DISMISSED** [review]
r=me, want shane's approval too

File: provider/datagen/src/transform/segmenter/mod.rs
Comment (sffc) [inline provider/datagen/src/transform/segmenter/mod.rs:167]: Issue: This is a complicated change and it is more difficult to review when you move around code that is unrelated to the change. It appears that you changed this function from being a module function to being an associated function.

Comment (sffc) [inline provider/datagen/src/transform/segmenter/mod.rs:167]: Issue: This is a complicated PR, and it is more difficult to review when you refactor code in a way that is unrelated to the change. It appears that you changed this function from being a module function to being an inline function. It would be a much smaller change if you would add `cfg`s to the module functions if that's all you need to make the code compile without warnings.

I didn't check if you have this in a standalone commit; please use more detailed commit messages if you prefer a commit-by-commit review.

**Review by sffc — COMMENTED** [review]
I approve of the change, but I encourage you next time to put more priority on making it easy for the code reviewer.

File: provider/datagen/src/bin/datagen/mod.rs
Comment (robertbastian) [reply provider/datagen/src/bin/datagen/mod.rs:16]: Well it's `pub mod` to the parent which doesn't reexport it, which we also do in lib crates for private modules.

File: provider/datagen/src/transform/segmenter/mod.rs
Comment (robertbastian) [reply provider/datagen/src/transform/segmenter/mod.rs:167]: This is the most recent commit called `seg cleanup`. I would have preferred to make this a standalone PR, but with the cross-timezone review cycle having chains is a massive pain.

**Review by Manishearth — APPROVED** [review]

File: provider/datagen/src/transform/segmenter/mod.rs
Comment (sffc) [reply provider/datagen/src/transform/segmenter/mod.rs:167]: OK thanks for the pointer; everything looks mergeable now except for the `seg cleanup` commit which requires more of my time so I will look at in more detail this afternoon.

If your goal is to get things reviewed faster, please don't scope-creep them in the middle of a review cycle.

Comment (Manishearth) [reply provider/datagen/src/transform/segmenter/mod.rs:167]: @robertbastian highly recommend just making a separate branch and creating a draft PR with "based on <other PR>, start reviewing at <commit name>"

@sffc I think given that this is a large PR already we should tend towards merging sooner rather than later and deal with the leftover stuff as post-review.

Comment (robertbastian) [reply provider/datagen/src/transform/segmenter/mod.rs:167]: > @robertbastian highly recommend just making a separate branch and creating a draft PR with "based on , start reviewing at "

This has not worked well for me in the past, I would get approval it but it has merge conflicts, I lose approval and it takes another couple of days to get approval again.

Comment (Manishearth) [reply provider/datagen/src/transform/segmenter/mod.rs:167]: I find rebases to be better than merges for dealing with that. But yeah that can be a pain.

Comment (sffc) [reply provider/datagen/src/transform/segmenter/mod.rs:167]: My preferred solution is to make a new PR as Draft, and once the parent PR is merged, fix up the child Draft PR and then send it out for review. Avoid PR chains longer than 2 or 3 by working on some other component that is independent; the project is big enough that there are lots of options.

**Review by Manishearth — APPROVED** [review]

**Review by sffc — APPROVED** [review]

**Pull request merged**
